panic: invalid specification "~>1.16": a space separator is required after the operator "~>"

### Terraform Version

```
0.13.0-beta1
```

### Terraform Configuration Files


```hcl
terraform {
  required_version = "~> 0.13.0"
#  experiments = [variable_validation]
}

provider "vsphere" {
  user           = var.vCentreDC.vCentreUser
  password       = var.vCentreDC.vCentrePassword
  vsphere_server = var.vCentreDC.vCentreAddress

  # If you have a self-signed cert
  allow_unverified_ssl = true
  version = "~>1.16"
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output

```
Initializing the backend...
panic: invalid specification "~>1.16": a space separator is required after the operator "~>"

goroutine 1 [running]:
github.com/hashicorp/terraform/internal/getproviders.MustParseVersionConstraints(0xc00012c10a, 0x6, 0x1, 0xc00012c10a, 0x6)
	/home/circleci/project/project/internal/getproviders/types.go:96 +0x80
github.com/hashicorp/terraform/configs.(*Config).addProviderRequirements(0xc000501520, 0xc00011a9f0, 0x100, 0xc000030000, 0x100)
	/home/circleci/project/project/configs/config.go:233 +0x8d1
github.com/hashicorp/terraform/configs.(*Config).ProviderRequirements(...)
	/home/circleci/project/project/configs/config.go:177
github.com/hashicorp/terraform/command.(*InitCommand).getProviders(0xc00014f340, 0xc000501520, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/circleci/project/project/command/init.go:416 +0x77
github.com/hashicorp/terraform/command.(*InitCommand).Run(0xc00014f340, 0xc00003a1a0, 0x1, 0x1, 0xc00007ee40)
	/home/circleci/project/project/command/init.go:282 +0x12f7
github.com/mitchellh/cli.(*CLI).Run(0xc00003c8c0, 0xc00003c8c0, 0xc00063fd48, 0x1)
	/home/circleci/project/project/vendor/github.com/mitchellh/cli/cli.go:255 +0x1da
main.wrappedMain(0x0)
	/home/circleci/project/project/main.go:275 +0xc90
main.realMain(0x0)
	/home/circleci/project/project/main.go:106 +0xde
main.main()
	/home/circleci/project/project/main.go:42 +0x3a
```

### Expected Behavior
Shouldn't crash. v.0.12.x accepts the version without the space.

### Actual Behavior
Terraform crashed

### Steps to Reproduce

1. `terraform init` with the above config

--- a/command/testdata/init-get-providers/main.tf
+++ b/command/testdata/init-get-providers/main.tf
@@ -7,5 +7,8 @@ provider "greater-than" {
 }
 
 provider "between" {
-  version = "> 1.0.0 , < 3.0.0"
+  # The second constraint here intentionally has
+  # no space after the < operator to make sure
+  # that we can parse that form too.
+  version = "> 1.0.0 , <3.0.0"
 }

--- a/configs/config.go
+++ b/configs/config.go
@@ -236,11 +236,22 @@ func (c *Config) addProviderRequirements(reqs getproviders.Requirements) hcl.Dia
 		}
 		// The model of version constraints in this package is still the
 		// old one using a different upstream module to represent versions,
-		// so we'll need to shim that out here for now. We assume this
-		// will always succeed because these constraints already succeeded
-		// parsing with the other constraint parser, which uses the same
-		// syntax.
-		constraints := getproviders.MustParseVersionConstraints(providerReqs.Requirement.Required.String())
+		// so we'll need to shim that out here for now. The two parsers
+		// don't exactly agree in practice 🙄 so this might produce new errors.
+		// TODO: Use the new parser throughout this package so we can get the
+		// better error messages it produces in more situations.
+		constraints, err := getproviders.ParseVersionConstraints(providerReqs.Requirement.Required.String())
+		if err != nil {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid version constraint",
+				// The errors returned by ParseVersionConstraint already include
+				// the section of input that was incorrect, so we don't need to
+				// include that here.
+				Detail:  fmt.Sprintf("Incorrect version constraint syntax: %s.", err.Error()),
+				Subject: providerReqs.Requirement.DeclRange.Ptr(),
+			})
+		}
 		reqs[fqn] = append(reqs[fqn], constraints...)
 	}
 	// Each resource in the configuration creates an *implicit* provider
@@ -272,7 +283,24 @@ func (c *Config) addProviderRequirements(reqs getproviders.Requirements) hcl.Dia
 			reqs[fqn] = nil
 		}
 		if provider.Version.Required != nil {
-			constraints := getproviders.MustParseVersionConstraints(provider.Version.Required.String())
+			// The model of version constraints in this package is still the
+			// old one using a different upstream module to represent versions,
+			// so we'll need to shim that out here for now. The two parsers
+			// don't exactly agree in practice 🙄 so this might produce new errors.
+			// TODO: Use the new parser throughout this package so we can get the
+			// better error messages it produces in more situations.
+			constraints, err := getproviders.ParseVersionConstraints(provider.Version.Required.String())
+			if err != nil {
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid version constraint",
+					// The errors returned by ParseVersionConstraint already include
+					// the section of input that was incorrect, so we don't need to
+					// include that here.
+					Detail:  fmt.Sprintf("Incorrect version constraint syntax: %s.", err.Error()),
+					Subject: provider.Version.DeclRange.Ptr(),
+				})
+			}
 			reqs[fqn] = append(reqs[fqn], constraints...)
 		}
 	}

--- a/go.mod
+++ b/go.mod
@@ -13,7 +13,7 @@ require (
 	github.com/apparentlymart/go-cidr v1.0.1
 	github.com/apparentlymart/go-dump v0.0.0-20190214190832-042adf3cf4a0
 	github.com/apparentlymart/go-userdirs v0.0.0-20190512014041-4a23807e62b9
-	github.com/apparentlymart/go-versions v0.0.2-0.20180815153302-64b99f7cb171
+	github.com/apparentlymart/go-versions v1.0.0
 	github.com/armon/circbuf v0.0.0-20190214190532-5111143e8da2
 	github.com/armon/go-metrics v0.0.0-20180917152333-f0300d1749da // indirect
 	github.com/armon/go-radix v1.0.0 // indirect

